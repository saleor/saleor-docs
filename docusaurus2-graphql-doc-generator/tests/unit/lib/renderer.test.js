const mock = require("mock-fs");
const path = require("path");
const dirTree = require("directory-tree");

const Renderer = require("@/lib/renderer");

jest.mock("@/lib/printer");
const Printer = require("@/lib/printer");

const FOLDER = "output";
const HOMEPAGE = "generated.md";

describe("lib", () => {
  describe("renderer", () => {
    describe("class Renderer", () => {
      let rendererInstance,
        baseURL = "graphql",
        printerInstance = new Printer("SCHEMA", baseURL, "root");

      beforeEach(() => {
        mock({
          node_modules: mock.load(
            path.resolve(__dirname, "../../../node_modules"),
          ),
          output: {},
          assets: {
            [HOMEPAGE]: mock.load(require.resolve(`@assets/${HOMEPAGE}`)),
          },
        });
        rendererInstance = new Renderer(printerInstance, FOLDER, baseURL);
      });

      afterEach(() => {
        mock.restore();
      });

      describe("renderTypeEntities()", () => {
        test("creates entity page into output folder", async () => {
          expect.assertions(2);
          jest
            .spyOn(printerInstance, "printType")
            .mockReturnValue("Lorem ipsum");
          const output = `${FOLDER}/foobar`;

          const meta = await rendererInstance.renderTypeEntities(
            output,
            "FooBar",
            "FooBar",
          );
          expect(meta).toEqual({ category: "Foobar", slug: "foobar/foo-bar" });

          expect(dirTree(FOLDER)).toMatchSnapshot();
        });
      });

      describe("renderSidebar()", () => {
        test("creates Docusaurus compatible sidebar.js into output folder", async () => {
          expect.assertions(1);

          const pages = [
            { category: "foo", slug: "lorem" },
            { category: "foo", slug: "ipsum" },
            { category: "bar", slug: "dolor" },
          ];

          await rendererInstance.renderSidebar(pages);
          expect(dirTree(FOLDER)).toMatchSnapshot();
        });
      });

      describe("generateSidebar()", () => {
        test("generate Docusaurus sidebar object from list of pages", async () => {
          const pages = [
            { category: "foo", slug: "lorem" },
            { category: "foo", slug: "ipsum" },
            { category: "bar", slug: "dolor" },
          ];
          const sidebar = rendererInstance.generateSidebar(pages);
          expect(sidebar).toMatchSnapshot();
        });
      });

      describe("renderHomepage()", () => {
        test("copies default homepage into output folder", async () => {
          expect.assertions(1);
          await rendererInstance.renderHomepage(`assets/${HOMEPAGE}`);

          expect(dirTree(FOLDER)).toMatchSnapshot({
            children: [{ size: expect.any(Number) }],
            size: expect.any(Number),
          });
        });
      });

      describe("renderRootTypes()", () => {
        test("render root type", async () => {
          expect.assertions(1);
          await rendererInstance.renderRootTypes("Object", [
            { name: "foo" },
            { name: "bar" },
          ]);
          expect(dirTree(FOLDER)).toMatchSnapshot();
        });
      });
    });
  });
});
