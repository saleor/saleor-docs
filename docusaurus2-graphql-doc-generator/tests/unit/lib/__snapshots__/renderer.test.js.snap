// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib renderer class Renderer generateSidebar() generate Docusaurus sidebar object from list of pages 1`] = `
Array [
  Object {
    "id": "graphql/schema",
    "type": "doc",
  },
  Object {
    "items": Array [
      "graphql/ipsum",
      "graphql/lorem",
    ],
    "label": "foo",
    "type": "category",
  },
  Object {
    "items": Array [
      "graphql/dolor",
    ],
    "label": "bar",
    "type": "category",
  },
]
`;

exports[`lib renderer class Renderer renderHomepage() copies default homepage into output folder 1`] = `
Object {
  "children": Array [
    Object {
      "extension": ".md",
      "name": "generated.md",
      "path": "output/generated.md",
      "size": Any<Number>,
      "type": "file",
    },
  ],
  "name": "output",
  "path": "output",
  "size": Any<Number>,
  "type": "directory",
}
`;

exports[`lib renderer class Renderer renderRootTypes() render root type 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "extension": ".mdx",
          "name": "bar.mdx",
          "path": "output/object/bar.mdx",
          "size": 11,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "foo.mdx",
          "path": "output/object/foo.mdx",
          "size": 11,
          "type": "file",
        },
      ],
      "name": "object",
      "path": "output/object",
      "size": 22,
      "type": "directory",
    },
  ],
  "name": "output",
  "path": "output",
  "size": 22,
  "type": "directory",
}
`;

exports[`lib renderer class Renderer renderSidebar() creates Docusaurus compatible sidebar.js into output folder 1`] = `
Object {
  "children": Array [
    Object {
      "extension": ".js",
      "name": "sidebar-schema.js",
      "path": "output/sidebar-schema.js",
      "size": 261,
      "type": "file",
    },
  ],
  "name": "output",
  "path": "output",
  "size": 261,
  "type": "directory",
}
`;

exports[`lib renderer class Renderer renderTypeEntities() creates entity page into output folder 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "extension": ".mdx",
          "name": "foo-bar.mdx",
          "path": "output/foobar/foo-bar.mdx",
          "size": 11,
          "type": "file",
        },
      ],
      "name": "foobar",
      "path": "output/foobar",
      "size": 11,
      "type": "directory",
    },
  ],
  "name": "output",
  "path": "output",
  "size": 11,
  "type": "directory",
}
`;
